added gocd functions around pipeline reading and sending to GoCD

frank0631 opened the pull request.

Commit:
gocd pipeline config json Marshal

Commit:
added functions to GET and POST to a gocd server

Commit:
Merge branch 'feature/gocd' into develop

marcesher commented on gocd/test.json:
```diff
@@ -0,0 +1,257 @@
+    {
+      "type": "git",
+      "attributes": {
+        "url": "https://github.cfpb.gov/databases/test.git",
```
Probably shouldn't have internal URLs / IPs in a public repo

tanderegg replied:
+1 @frank0631 can you replace that with a public url?  Something like https://github.com/cfpb/rhobot-test.git?

frank0631 replied:
ip's and cfpb internal url's have been replaced

marcesher commented:
Please spend some time on the README, so that it's clearer what this thing even is. I understand that it's just getting started; still, we should practice good hygiene

tanderegg commented:
Also +1, @frank0631 just provide a brief description of Rhobot, feel free to pull directly from the design document.

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,177 @@
+
+	payloadBytes, err := json.Marshal(pipelineConfig)
+	if err != nil {
+		// handle err
```
@frank0631 I'd prefer not to commit code that does no error handling at all, could you look into providing some baseline handling of these cases?  It probably would make the most sense for this function to return a Pipeline and an Error, in which case you'd just return (nil, err), but you could also panic with the error message for now if you wish.  See: https://golang.org/doc/effective_go.html#errors

frank0631 replied:
errors now bubble up from functions in the gocd pkg. An example of catching the errors can be see in the gocd_test.go

Commit:
errors bubble up from gocd pkg

also removed references to internal cfpb ip's

Commit:
gocd pkg

make Pull/Push/Clone & ReadPipelineJSONFromFile public. Load GoCD user &
pass from env variables

Commit:
add cli for GoCD interaction

frank0631 commented:
The README and other non-code files that need to be updated will be in a seperate pull request from this pull request: https://github.com/cfpb/rhobot/pull/3

tanderegg committed:
Minor fixes to make the build process work, and suggestions from gofmt

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+	return unmarshalPipeline(body)
+}
+
+func Push(gocdULR string, path string, group string) {
```
Go likes to have all public functions contain a line of comment, starting with the function name, like // Push ....
We should follow that convention.

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+	return unmarshalPipeline(data)
+}
+
+func pipelineConfigPOST(gocdULR string, pipelineConfig PipelineConfig) (pipeline Pipeline, err error) {
```
gocdURL not gocdULR

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+	}
+	payloadBody := bytes.NewReader(payloadBytes)
+
+	req, err := http.NewRequest("POST", gocdULR+"/go/api/admin/pipelines", payloadBody)
```
gocdURL

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+	if err != nil {
+		return pipeline, err
+	}
+	//fmt.Println("pipelineConfigPOST:", string(prettyJSON.Bytes()))
```
Preferable to using comments, we should be able to run in a formal "debug" or "verbose" mode.  Look into Logrus or  glog?

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+	return unmarshalPipeline(body)
+}
+
+func pipelineGET(gocdULR string, pipelineName string) (pipeline Pipeline, err error) {
```
gocdURL

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+
+	// Generated by curl-to-Go: https://mholt.github.io/curl-to-go
+
+	req, err := http.NewRequest("GET", gocdULR+"/go/api/admin/pipelines/"+pipelineName, nil)
```
gocdURL

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+	if err != nil {
+		return pipeline, err
+	}
+	//fmt.Println("pipelineConfigGET:", string(prettyJSON.Bytes()))
```
See above comment by logging

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+	pipeline, err := ReadPipelineJSONFromFile(path)
+	check(err)
+	pipelineConfig := PipelineConfig{group, pipeline}
+	_, err = pipelineConfigPOST(gocdULR, pipelineConfig)
```
gocdURL

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+	check(err)
+}
+
+func Pull(gocdULR string, path string) {
```
gocdURL

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+	check(err)
+	name := pipeline.Name
+
+	Clone(gocdULR , path , name )
```
gocdURL

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+	Clone(gocdULR , path , name )
+}
+
+func Clone(gocdULR string, path string, name string) {
```
gocdURL

tanderegg commented on gocd/gocd.go:
```diff
@@ -0,0 +1,210 @@
+
+func Clone(gocdULR string, path string, name string) {
+
+	pipelineFetched, err := pipelineGET(gocdULR, name)
```
gocdURL

tanderegg commented on gocd/gocd_test.go:
```diff
@@ -0,0 +1,70 @@
+	"bytes"
+	"encoding/json"
+	"fmt"
+	//"github.com/davecgh/go-spew/spew"
```
Let's remove commented spew code for now

tanderegg commented on gocd/test.json:
```diff
@@ -0,0 +1,206 @@
+        {
+            "type": "git",
+            "attributes": {
+                "url": "https://github.com/cfpb/rhobot-test.git?",
```
Remove unneeded "?"

frank0631 replied:
the "?" was in the test pipeline in GoCD. I've modified the pipeline and pulled it

frank0631 committed:
Merge pull request #1 from tanderegg/build_fixes

Minor fixes to make the build process work, and suggestions from gofmt

frank0631 committed:
gofmt and golint compliant

frank0631 committed:
removed spew

frank0631 committed:
edited gocd pipeline test's source material

tanderegg commented:
Looks good, thanks @frank0631!
